### PR TITLE
Oil painting effect

### PR DESCRIPTION
**Pull request by LaurentBerger**
Oil painting effect 

image

Ref http://spinroot.com/pico/a4.pdf page 46-47

### CODE REVIEW

**Comment by alalek** [comment]
Use `cvRound`:
```
cvRound(7.0 / 2.0) => 4
```

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*M///////////////////////////////////////////////////////////////////////////////////////
+//
```
Comment (alalek) [inline modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: Please use short license header:
```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.
```

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/** @brief oilPainting
+See the book @cite Holzmann1988 for details.
+@param _src Input three-channel or one channel image (either CV_8UC3 or CV_8UC1)
+@param _dst Output image of the same size and type as src.
+@param size neighbouring size is 2-size+1
+@param dynRatio image is divided by dynRatio before histogram processing
+@param code	color space conversion code(see ColorConversionCodes). Histogram will used only first plane
+*/
+    CV_EXPORTS_W void oilPainting(InputArray _src, OutputArray _dst, int size, int dynRatio, int code);
```
Comment (alalek) [inline modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: > _src

In declarations (header) there is no necessary to add underscore prefix. Removing them will make documentation better.

File: modules/xphoto/src/oilpainting.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    void oilPainting(InputArray _src, OutputArray _dst, int size, int dynValue)
+    {
+        oilPainting(_src, _dst, size, dynValue, COLOR_BGR2GRAY);
+    }
+
+    void oilPainting(InputArray _src, OutputArray _dst, int size, int dynValue,int code)
+    {
+        CV_Assert(_src.kind() == _InputArray::MAT && size>=1 && dynValue>0 && dynValue<128);
+        Mat src = _src.getMat();
+        CV_Assert(src.type()==CV_8UC1 || src.type() == CV_8UC3);
```
Comment (alalek) [inline modules/xphoto/src/oilpainting.cpp]: Please use `CV_CheckTypeEQ(src.type(), src.type()==CV_8UC1 || src.type() == CV_8UC3, "")` to provide useful messages for OpenCV users.

Assertion above should be spitted too. "int" values should be validated by `CV_Check*()` macros.

File: modules/xphoto/src/oilpainting.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        {
+            cvtColor(src, lum, code);
+            if (lum.channels() > 1)
+            {
+                extractChannel(lum, lum, 0);
+            }
+        }
+        else
+            lum = src.clone();
+        lum.forEach<uchar>(Quotient(dynValue));
```
Comment (alalek) [inline modules/xphoto/src/oilpainting.cpp]: You can use C++11 lambdas here:

```
lum.forEach<uchar>([=](uchar &pixel, const int * /*position*/) { pixel = pixel / dynValue; });
```

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp (hunk 1 above)
Comment (LaurentBerger) [reply modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: OK

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp (hunk 2 above)
Comment (LaurentBerger) [reply modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: OK

File: modules/xphoto/src/oilpainting.cpp (hunk 4 above)
Comment (LaurentBerger) [reply modules/xphoto/src/oilpainting.cpp]: OK

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp (hunk 1 above)
Comment (alalek) [reply modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: Please update this one.

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+namespace cv
+{
+namespace xphoto
+{
+
+//! @addtogroup xphoto
+//! @{
+
+    /** @brief oilPainting
+    See the book @cite Holzmann1988 for details.
```
Comment (alalek) [inline modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: Please avoid indentation in namespaces.

File: modules/xphoto/src/oilpainting.cpp [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+
+
+namespace cv
+{
+namespace xphoto
+{   template<typename Type>
+    class ParallelOilPainting : public ParallelLoopBody
+    {
```
Comment (alalek) [inline modules/xphoto/src/oilpainting.cpp]: Please avoid indentation in namespaces.

File: modules/xphoto/include/opencv2/xphoto/oilpainting.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+
+#ifndef __OPENCV_OIL_PAINTING_HPP__
+#define __OPENCV_OIL_PAINTING_HPP__
+
+/** @file
+@date Sept 26, 2018
```
Comment (catree) [inline modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: Time travel? :smiley:

Comment (LaurentBerger) [reply modules/xphoto/include/opencv2/xphoto/oilpainting.hpp]: oops!

**Pull request merged**